in-progress/0003-native-merkle-trees.md

```diff
@@ -0,0 +1,212 @@
+|                      |                                                                                               |
+| -------------------- | --------------------------------------------------------------------------------------------- |
+| Owners               | @PhilWindle                                                                                   |
+| Approvers            | @just-mitch @alexghr                                                                          |
+| Target Approval Date | 2024-06-21                                                                                    |
+
+
+## Executive Summary
+
+This design attempts to solve the problem of slow sync and merkle tree insertion performance.
+
+
+## Introduction
+
+We require high performance merkle tree implementations both to ensure nodes can stay synched to the network and sequencers/provers can advance the state as required to build blocks. Our cuirrent TS implementations are limited in their single-threaded nature and the unavoidable constraint of have to repeatedly call into WASM to perform a hash operation.
+
+Some analysis of the quantity of hashing and the time required can be found [here](https://hackmd.io/@aztec-network/HyfTK9U5a?type=view).
+
+This design proposes the creation of a set of multi-threaded merkle tree implementations in C++ using LMDB. It builds upon some previous prototyping to develop concurrent indexed tree insertions.
+
+## Implementation
+
+There are many parts to this design, we will walk through them individiually and discuss the choices made at each stage.
+
+### Overall Architecture
+
+A new C++ binary, World State, will be created that will be started by the node software. It will be configured with the location in which Merkle Tree data should be stored. It will then accept and respond with msgpack-ed messages over one or more streams. The initial implementation will simply used stdio, but this will be absrtacted such that this could be replaced by other stream-based mechanisms.
```

Comment:
In the prototype branch the source for the tree lives in barretenberg. Is that where you plan to keep it? This feels like it depends on barretenberg, but is not part of it.

Reply:
As I understand it, bb is structured that everything lives under `barretenberg` but anything not proving system related sits under `crypto`. Most of the work I have done builds on some tree-related code that was already under `crypto/merkle-tree`.